qp_klp/Step.py

```diff
@@ -326,23 +328,134 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
+                    f'{LOG_PREFIX}-{PREPFILE_JOB}.{TAR_EXT}', 'OUTPUT_FIRST'),
+
+                   (['GenPrepFileJob/PrepFiles'], TAR_CMD,
+                    f'{PREP_PREFIX}.{TAR_EXT}', 'OUTPUT_FIRST'),
+
+                   (['failed_samples.html', 'touched_studies.html'],
+                    'mv', RESULTS_DIR, 'INPUTS_FIRST'),
+
+                   (['FastQCJob/multiqc'], 'mv', RESULTS_DIR, 'INPUTS_FIRST')]
+
+        cmds = []
+
+        for inputs, action, output, order in op_meta:
+            confirmed_inputs = []
+            for input in inputs:
+                if exists(join(self.pipeline.output_path, input)):
+                    # it's expected that some inputs may not exist due to
+                    # different pipeline types. If one or more inputs do not
+                    # exist, do not include them in the command-line as they
+                    # may cause an error.
+                    confirmed_inputs.append(input)
+
+            # do not add the command to the list unless at least one of
+            # the inputs exists. It's okay for a command to go unprocessed.
+            if confirmed_inputs:
+                # convert to string form before using.
+                confirmed_inputs = ' '.join(confirmed_inputs)
+                if order == 'OUTPUT_FIRST':
+                    cmds.append(f'{action} {output} {confirmed_inputs}')
+                elif order == 'INPUTS_FIRST':
+                    cmds.append(f'{action} {confirmed_inputs} {output}')
+                else:
+                    raise ValueError(f"'{order}' is not a defined order of "
+                                     "operations")
+
+        return cmds
+
+    def generate_commands(self):
+        cmds = self._helper_process_operations()
 
-        cmds = ['tar zcvf logs-ConvertJob.tgz ConvertJob/logs',
-                'tar zcvf logs-FastQCJob.tgz FastQCJob/logs',
-                'tar zcvf reports-FastQCJob.tgz FastQCJob/fastqc',
-                'tar zcvf logs-GenPrepFileJob.tgz GenPrepFileJob/logs',
-                'tar zcvf prep-files.tgz GenPrepFileJob/PrepFiles']
-        self.cmds = [f'cd {out_dir}; {x}' for x in cmds]
+        result = self._helper_process_fastp_report_dirs()
+
+        if result:
+            cmds.append(result)
+
+        cmds.append(self._helper_process_blanks())
+
+        # if one or more tar-gzip files are found (which we expect there to
+        # be), move them into the 'final_results' directory. However, if none
+        # are present, don't raise an error.
+        cmds.append('(find *.tgz -maxdepth 1 -type f | xargs mv -t '
+                    'final_results) || true')
+
+        # confirm uploads files exist as expected before using helper
+        # methods.
+        for uploads_folder in [proj[1] for proj in self.special_map]:
+            if not exists(uploads_folder):
+                raise ValueError(f"Uploads folder '{uploads_folder}' does "
+                                 "not exist")
```

Comment:
If nothing is gonna be uploaded, do we still need this?

Reply:
fair point. :)